File: Python/Companion.py
from FriendlyNPC import FriendlyNPC
from CanBattle import CanBattle

class Companion(FriendlyNPC, CanBattle):
    """FriendlyNPC yang bisa ikut bertarung dengan player."""
    def __init__(self, name: str, gender: str, race: str, affiliation,
                 maxAP: int, maxHP: int, maxM: int, charClass: str,
                 is_marryable: bool, birthday: str, haveQuest,
                 fee: int, grade: str):
        
        # Panggil konstruktor kelas berikutnya dalam MRO, 
        # meneruskan SEMUA argumen yang dibutuhkan oleh FriendlyNPC dan CanBattle.
        # Python MRO akan memastikan setiap base class mendapatkan argumen yang tepat.
        super().__init__(name=name, gender=gender, race=race, affiliation=affiliation,
                         is_marryable=is_marryable, birthday=birthday, haveQuest=haveQuest,
                         maxAP=maxAP, maxHP=maxHP, maxM=maxM, charClass=charClass)
                         
        self.fee = fee
        self.grade = grade

    def display(self):
        # Memanggil display secara eksplisit untuk menghindari duplikasi display Karakter
        FriendlyNPC.display(self) 
        
        # Tampilkan detail CanBattle yang unik (sudah ter-init di super())
        print(f"  Kelas (Battle): {self.charClass}")
        print(f"  Max HP: {self.maxHealth}")
        print(f"  Max Mana: {self.maxMana}")
        print(f"  Max AP: {self.maxActionPoint}")
        
        # Detail Companion
        print(f"  Bayaran: {self.fee}")
        print(f"  Peringkat: {self.grade}")

File: C++/Companion.cpp
#pragma once
#include "FriendlyNPC.cpp"
#include "CanBattle.cpp"

class Companion : public CanBattle, public FriendlyNPC {
private:
    int fee;
    std::string grade; // F/E/D/C/B/A/S/SS

public:
    Companion(std::string name, std::string gender, std::string race, Factions* affiliation,
              int maxAP, int maxHP, int maxM, std::string charClass,
              bool is_marryable, std::string birthday, Quest* quest,
              int fee, std::string grade)
        : Karakter(name, gender, race, affiliation), // <-- Companion (most derived) initializes VIRTUAL BASE Karakter
          CanBattle(name, gender, race, affiliation, maxAP, maxHP, maxM, charClass),
          FriendlyNPC(name, gender, race, affiliation, is_marryable, birthday, quest),
          fee(fee), grade(grade) {}

    // Getter
    int getFee() const { return fee; }
    std::string getGrade() const { return grade; }

    // Setter
    void setFee(int f) { fee = f; }
    void setGrade(std::string g) { grade = g; }

    // Override Display
    void display() const override {
        // Cukup panggil salah satu base class yang menampilkan Karakter (misal FriendlyNPC)
        FriendlyNPC::display();
        
        // Menambahkan detail CanBattle yang hilang (HP/AP/Mana)
        std::cout << "  Kelas (Battle): " << CanBattle::getCharClass() << std::endl;
        std::cout << "  Max HP: " << CanBattle::getMaxHealth() << std::endl;
        std::cout << "  Max Mana: " << CanBattle::getMaxMana() << std::endl;
        std::cout << "  Max AP: " << CanBattle::getMaxActionPoint() << std::endl;

        // Detail Companion
        std::cout << "  Bayaran: " << fee << std::endl;
        std::cout << "  Peringkat: " << grade << std::endl;
    }
};